### PR TITLE
Use modern cmake

### PR DESCRIPTION
**Pull request by schweitzer**
Hello guys,

First of all thanks for your work, openVSLAM is really great.

In order to use openVSLAM as a library, I've made a lot of changes (mainly focused on CMake), I would like the share it with you because I think this could be interesting for the community.

* Modernize CMake usage, by using imported target when available, removing unnecessary c++11 checks.
* Add install rules on each openvslam modules to be able to  launch a `make install`.
* Add rules to generate `openvslamTargets.cmake` with exported targets, so that openvslam can be used as a CMake package.

This is only working on Linux (maybe Apple also) for now (compilation on windows leads to some issue already referenced [here](https://github.com/xdspacelab/openvslam/issues/108)).

I'm able to compile, install and run openvslam from an outside application.

Unfortunately I did't test all kind of openvslam usage (ROS, SocketViewer, ...)

Please let me know if there is anything wrong with the PR.

### CODE REVIEW

**Comment by shinsumicco** [comment]
@schweitzer
Thank you so much!
Your PR is very helpful for us because we haven't had much time to spend on refactoring cmake configurations.
I'll review it within a few days.

File: CMakeLists.txt [hunk 1]
```diff
@@ -98,129 +101,70 @@ endif()
...
-# json
-set(json_INCLUDE_DIR ${PROJECT_SOURCE_DIR}/3rd/json/include)
-include_directories(${json_INCLUDE_DIR})
+find_package(yaml-cpp REQUIRED)
 
 # pangolin viewer
 set(USE_PANGOLIN_VIEWER ON CACHE BOOL "Enable Pangolin Viewer")
 if(USE_PANGOLIN_VIEWER)
-    find_package(Pangolin REQUIRED)
+    find_package(Pangolin QUIET REQUIRED)
```
Comment (shinsumicco) [inline CMakeLists.txt]: Could you tell me why you insert `QUIET` modifier for this?

File: .gitignore [hunk 2]
```diff
@@ -38,3 +38,5 @@ cmake-build-release/
 *.exe
 *.out
 *.app
+
+*.user
```
Comment (shinsumicco) [inline .gitignore]: Is this for Visual Studio?

**Comment by shinsumicco** [comment]
I haven't finished the review yet, but have some questions.
Could you answer the following questions?
- compiler and cmake version you use
- the reason why you insert `QUIET` to some `find_package()`
- compilation results of test codes (succeeded or failed)

And, could you align indents of some source codes?
`CMakeLists.txt` are currently written with four-space indent.

Thanks.

**Comment by schweitzer** [comment]
> I haven't finished the review yet, but have some questions.
> Could you answer the following questions?
> 
> * compiler and cmake version you use
> * the reason why you insert `QUIET` to some `find_package()`
> * compilation results of test codes (succeeded or failed)
> 
> And, could you align indents in some source codes?
> `CMakeLists.txt` are currently written with four-space indent.
> 
> Thanks.

Hello and thank you for your review, 

* I'm currently using **g++ 7.4** along with cmake **3.15.3**, but I've also tested with g++ version **8.3.0**

* The `QUIET` shouldn't be here, usually I'm using it to disable message when package cannot be found, it can be useful when developing and doing some tests with cmake, but I think I should remove it from "final" version.

* I've forget to update the `CMakeLists.txt` from tests folder, I think the link with `Eigen::Eigen3` is missing.

* Concerning the indentation issue, can you point me a wrong indentation you want me to fix? I'm also using 4 space as indent :thinking: 

I will provide you a commit that fix the above points as soon as possible.

Thanks again !

**Comment by schweitzer** [comment]
With my last commits, I'm able to compile and run the tests (`nj` = ninja)

```shell
mschweitzer@IRCW123:~/dev/others/build/openvslam/release$ nj test
[0/1]   0% 0.000s Running tests...
Test project /home/mschweitzer/dev/others/build/openvslam/release
      Start  1: test_openvslam_feature_orb_extractor
 1/10 Test  #1: test_openvslam_feature_orb_extractor ......   Passed    0.38 sec
      Start  2: test_openvslam_feature_orb_params
 2/10 Test  #2: test_openvslam_feature_orb_params .........   Passed    0.01 sec
      Start  3: test_openvslam_match_angle_checker
 3/10 Test  #3: test_openvslam_match_angle_checker ........   Passed    0.00 sec
      Start  4: test_openvslam_match_base
 4/10 Test  #4: test_openvslam_match_base .................   Passed    0.00 sec
      Start  5: test_openvslam_solve_essential_solver
 5/10 Test  #5: test_openvslam_solve_essential_solver .....   Passed    0.02 sec
      Start  6: test_openvslam_solve_fundamental_solver
 6/10 Test  #6: test_openvslam_solve_fundamental_solver ...   Passed    0.02 sec
      Start  7: test_openvslam_solve_homography_solver
 7/10 Test  #7: test_openvslam_solve_homography_solver ....   Passed    0.02 sec
      Start  8: test_openvslam_util_fancy_index
 8/10 Test  #8: test_openvslam_util_fancy_index ...........   Passed    0.00 sec
      Start  9: test_openvslam_util_random_array
 9/10 Test  #9: test_openvslam_util_random_array ..........   Passed    0.01 sec
      Start 10: test_openvslam_util_trigonometric
10/10 Test #10: test_openvslam_util_trigonometric .........   Passed    0.00 sec

100% tests passed, 0 tests failed out of 10

Total Test time (real) =   0.47 sec

```

**Comment by shinsumicco** [comment]
Thank you for providing the information of your environment to me and confirming the build of test codes!!

Some of our team members reported the following compile errors.
Do you know what caused this?

- With Ubuntu 18.04 + gcc 7.4 + OpenCV 4.1.2
```
In file included from /anonymous/openvslam/test/helper/bearing_vector.h:4:0,
                 from /anonymous/openvslam/test/helper/bearing_vector.cc:1:
/anonymous/openvslam/src/openvslam/type.h:12:10: fatal error: opencv2/core/types.hpp: No such file or directory
 #include <opencv2/core/types.hpp>
          ^~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
test/helper/CMakeFiles/test_helper.dir/build.make:62: recipe for target 'test/helper/CMakeFiles/test_helper.dir/bearing_vector.cc.o' failed
make[2]: *** [test/helper/CMakeFiles/test_helper.dir/bearing_vector.cc.o] Error 1
CMakeFiles/Makefile2:1211: recipe for target 'test/helper/CMakeFiles/test_helper.dir/all' failed
make[1]: *** [test/helper/CMakeFiles/test_helper.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
```

- With macOS 10.15 + clang  7.0.0 + OpenCV 3.4.8
```
In file included from /anonymous/openvslam/test/helper/landmark.cc:1:
In file included from /anonymous/openvslam/test/helper/landmark.h:4:
/anonymous/openvslam/src/openvslam/type.h:12:10: fatal error: 'opencv2/core/types.hpp' file not found
#include <opencv2/core/types.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from /anonymous/openvslam/test/helper/keypoint.cc:1:
In file included from /anonymous/openvslam/test/helper/keypoint.h:4:
/anonymous/openvslam/src/openvslam/type.h:12:10: fatal error: 'opencv2/core/types.hpp' file not found
#include <opencv2/core/types.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from /anonymous/openvslam/test/helper/bearing_vector.cc:1:
In file included from /anonymous/openvslam/test/helper/bearing_vector.h:4:
/anonymous/openvslam/src/openvslam/type.h:12:10: fatal error: 'opencv2/core/types.hpp' file not found
#include <opencv2/core/types.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
1 error generated.
make[2]: *** [test/helper/CMakeFiles/test_helper.dir/keypoint.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [test/helper/CMakeFiles/test_helper.dir/landmark.cc.o] Error 1
1 error generated.
make[2]: *** [test/helper/CMakeFiles/test_helper.dir/bearing_vector.cc.o] Error 1
make[1]: *** [test/helper/CMakeFiles/test_helper.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

Thanks!

**Comment by shinsumicco** [comment]
The following codes are not four-space indent.

https://github.com/IRCAD-IHU/openvslam/blob/870e4cbb456c74a18c2de055368750e5e30401b1/src/socket_publisher/CMakeLists.txt#L28-L32

https://github.com/IRCAD-IHU/openvslam/blob/use-modern-cmake/src/openvslam/CMakeLists.txt#L140-L143

However, some configuration I'd wrote does not fit to the four-space indent style, too.
I'll reformat indents of all of the `CMakeLists.txt` just before merging this PR.
So, you don’t have to worry too much. Thanks!

File: test/helper/CMakeLists.txt [hunk 3]
```diff
@@ -5,4 +5,7 @@ add_library(test_helper SHARED
         landmark.cc)
 # add include directory as PUBLIC (because the headers are included in test codes)
 target_include_directories(test_helper SYSTEM
-        PUBLIC ${PROJECT_SOURCE_DIR}/test)
+        PUBLIC ${PROJECT_SOURCE_DIR}/test
+        ${PROJECT_SOURCE_DIR}/src)
+
+target_link_libraries(test_helper PUBLIC Eigen3::Eigen)
```
Comment (shinsumicco) [inline test/helper/CMakeLists.txt]: Please replace with
```cmake
target_link_libraries(test_helper PUBLIC Eigen3::Eigen opencv_core) 
```

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -33,8 +34,14 @@ foreach(SOURCE_PATH ${SOURCE_PATHS})
             test_helper
             gtest_main)
```
Comment (shinsumicco) [inline test/CMakeLists.txt]: Please replace with
```cmake
    target_link_libraries(${TEST_EXECUTABLE_NAME}
            ${PROJECT_NAME}
            test_helper
            gtest_main
            opencv_imgcodecs
            opencv_highgui)
```

File: src/pangolin_viewer/CMakeLists.txt [hunk 5]
```diff
@@ -1,14 +1,24 @@
...
 
 set_target_properties(pangolin_viewer PROPERTIES
         OUTPUT_NAME pangolin_viewer
         LIBRARY_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/lib)
 
-target_include_directories(pangolin_viewer PUBLIC
-        ${Pangolin_INCLUDE_DIRS})
+target_link_libraries(pangolin_viewer PUBLIC
+        ${PROJECT_NAME} 
+        pangolin)
```
Comment (shinsumicco) [inline src/pangolin_viewer/CMakeLists.txt]: Please replace with
```cmake
target_link_libraries(pangolin_viewer PUBLIC
        ${PROJECT_NAME}
        opencv_highgui
        pangolin)
```

**Review by shinsumicco — CHANGES_REQUESTED** [review]
I haven't finished reviewing all of the code yet, but I found some configuration to be modified in order to pass compile and link.
Could you check my comments and modify them accordingly if you agree to them?

Thanks again!

**Comment by schweitzer** [comment]
@shinsumicco : Thanks for the review and sorry for the late reply, I have a lot of work to do this week, so the corrections may take some time. But don't worry I'm going to do it :wink: !

Thanks again !

File: CMakeLists.txt [hunk 6]
```diff
@@ -98,124 +101,65 @@ endif()
...
-    remove_definitions(-DUSE_OPENMP)
-    message(STATUS "OpenMP: DISABLED")
-endif()
 
 # OpenCV
-find_package(OpenCV 3.3.1 QUIET)
+find_package(OpenCV 3.3.1 COMPONENTS core features2d calib3d)
 if(NOT OpenCV_FOUND)
-    find_package(OpenCV 4.0)
+    find_package(OpenCV 4.0 REQUIRED COMPONENTS core features2d calib3d)
```
[the remark this answers is not in the document]
Comment (schweitzer) [reply CMakeLists.txt]: Should be ok now.

File: test/helper/CMakeLists.txt (hunk 3 above)
Comment (schweitzer) [reply test/helper/CMakeLists.txt]: Should be ok now.

File: test/CMakeLists.txt (hunk 4 above)
Comment (schweitzer) [reply test/CMakeLists.txt]: Should be ok now.

File: src/pangolin_viewer/CMakeLists.txt (hunk 5 above)
Comment (schweitzer) [reply src/pangolin_viewer/CMakeLists.txt]: Should be ok now.

**Review by schweitzer — COMMENTED** [review]
I've pushed the modifications you ask, please let me know if it's ok for you.

File: CMakeLists.txt [hunk 7]
```diff
@@ -98,124 +101,65 @@ endif()
...
-        message(STATUS "OpenMP: DISABLED")
-    endif()
-else()
-    remove_definitions(-DUSE_OPENMP)
-    message(STATUS "OpenMP: DISABLED")
-endif()
 
 # OpenCV
-find_package(OpenCV 3.3.1 QUIET)
+find_package(OpenCV 3.3.1 COMPONENTS core features2d calib3d)
```
[the remark this answers is not in the document]
Comment (schweitzer) [reply CMakeLists.txt]: Should be ok now.

**Comment by furorenormannorum** [comment]
I just tested building the "modern cmake" proposal in Windows 10 64-bit, using Visual Studio 2017, and it might be a good idea include GNU related compiler flags only if GNUCXX is detected...e.g. do something similar to the following in the "Set compiler options" section:
if(CMAKE_COMPILER_IS_GNUCXX)
    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -Wextra")
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra")
endif(CMAKE_COMPILER_IS_GNUCXX)

**Comment by schweitzer** [comment]
@furorenormannorum 
> I just tested building the "modern cmake" proposal in Windows 10 64-bit, using Visual Studio 2017, and it might be a good idea include GNU related compiler flags only if GNUCXX is detected...e.g. do something similar to the following in the "Set compiler options" section:
> if(CMAKE_COMPILER_IS_GNUCXX)
> set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -Wextra")
> set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra")
> endif(CMAKE_COMPILER_IS_GNUCXX)

Yes you are right, in fact playing with CXX_FLAGS is never a good idea in CMake. 
Since Windows support isn't the purpose of this PR, I don't wanted to change this for now.

I can add the necessary changes if needed, but in my opinion this need to be integrated in a full windows support PR.

There are already some issues about windows, please see https://github.com/xdspacelab/openvslam/issues/108

**Comment by furorenormannorum** [comment]
As you yourself stated, modifying directly the CMAKE_CXX_FLAGS or CMAKE_C_FLAGS is not the "modern" CMake approach. Naturally there are some cases where this might be necessary, but since GCC is not the only C/C++ compiler that can be used in Linux, and I'm not sure if -Wall and -Wextra flags are known by compilers like Clang or Intel C++ compiler, I would still propose to make the CMake scripts as compiler agnostic as possible, if you have the time.

File: CMakeLists.txt [hunk 8]
```diff
@@ -241,57 +185,57 @@ endif()
...
+set(SOCKER_VIEWER_INCLUDE_INSTALL_DIR ${INCLUDES_DESTINATION}/socket_viewer)
 
 ###################
 # Build Selection #
 ###################
 
 add_subdirectory(src)
-add_subdirectory(example)
+
+set(BUILD_EXAMPLES OFF CACHE BOOL "Build Examples")
```
Comment (shinsumicco) [inline CMakeLists.txt]: Could you set BUILD_EXAMPLES as ON by default?
Because I think most users will be try the examples.

**Review by shinsumicco — COMMENTED** [review]
Thank you for making the modification. 
Could you check my comment and modify it accordingly if you agree to it?
https://github.com/xdspacelab/openvslam/pull/172/files#r353638441

Thanks again!

File: CMakeLists.txt (hunk 8 above)
Comment (furorenormannorum) [reply CMakeLists.txt]: I'm not the author of this patch, I just proposed to to make the CMake scripts compiler agnostic. However, I'll be more than glad to help where I can!

Comment (shinsumicco) [reply CMakeLists.txt]: Hi, @fabianschenk.
Of course I'd like to make CMake configuration compiler-independent especially for build on non-gcc compilers (e.g. clang) and MSVC.
But unfortunately, I cannot spend time on it at the moment owing to works for my master's course thesis.
So, we would be grateful if you make a new PR after this PR is merged.
Thank you for your contribution!

Comment (shinsumicco) [reply CMakeLists.txt]: Hi, @furorenormannorum.
Of course I'd like to make CMake configuration compiler-independent especially for build on non-gcc compilers (e.g. clang) and MSVC.
But unfortunately, I cannot spend time on it at the moment owing to works for my master's course thesis.
So, we would be grateful if you make a new PR after this PR is merged.
Thank you for your contribution!

Comment (furorenormannorum) [reply CMakeLists.txt]: Hi @shinsumicco, sure, no problem...I can make another PR after this one has been merged! Best of luck with the MsC thesis!!!

**Comment by shinsumicco** [comment]
Hello, @schweitzer.
I applied some modification to your PR and will merge it to `develop` because I have some spare time on this weekend 😉
Thank you for your helpful contribution!
And we would be grateful for your continued support!

File: CMakeLists.txt (hunk 8 above)
Comment (shinsumicco) [reply CMakeLists.txt]: Hi, @furorenormannorum.
This PR has just been merged.
We'd like to discuss your suggestion at the new PR.
Thanks!

**Comment by shinsumicco** [comment]
I've confirmed that CMake can find OpenVSLAM library with the following toy-example configuration:

```cmake
cmake_minimum_required(VERSION 3.1)
project(openvslam LANGUAGES CXX C)

set(CMAKE_CXX_STANDARD 11)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

find_package(openvslam REQUIRED)
add_executable(main main.cc)
target_link_libraries(main PRIVATE openvslam::openvslam)
```

**Pull request merged**